### PR TITLE
✨ Rebase to 0.19

### PR DESCRIPTION
**Pull request by tnthornton**
Rebased kcp-0.18 on top of kcp-0.19 (currently the branch from upstream).

This is currently needed for two reasons:
1. Pull in lastest changes from controller-runtime.
2. There was a signature change in utilruntime.PanicHandlers (0.31.0)

### CODE REVIEW

**Comment by tnthornton** [comment]
For some extra context, we started running into trouble with upgrading our k8s dependencies to 0.31.0+ due to the signature change in `PanicHandlers`. Happy to adjust any of the changes in here if I happened to miss a part of the process you all have going 👍

File: pkg/client/client.go [hunk 1]
```diff
@@ -71,6 +71,14 @@ type CacheOptions struct {
...
 type NewClientFunc func(config *rest.Config, options Options) (Client, error)
 
+// NewAPIReaderFunc allows a user to define how to create an API server reader.
+type NewAPIReaderFunc func(config *rest.Config, options Options) (Reader, error)
+
+// NewAPIReader creates a new API server reader.
+func NewAPIReader(config *rest.Config, options Options) (Reader, error) {
+	return New(config, options)
+}
+
```
Comment (sttts) [inline pkg/client/client.go:81]: in this file the last changes are missing, compare `KcpClusterDiscoveryCacheSize` and `MapperWithContext`.

Comment (sttts) [reply pkg/client/client.go:81]: https://github.com/kcp-dev/controller-runtime/pull/57/files

**Comment by sttts** [comment]
Compared to https://github.com/kcp-dev/controller-runtime/compare/12cc8d59fabe3a1dbbd927ae6aea077cf7962267..kcp-0.18.

File: pkg/client/client.go (hunk 1 above)
Comment (tnthornton) [reply pkg/client/client.go:81]: Weird, I wonder how I lost that 👀

**Pull request closed without merging**